src/app/loadout-builder/mod-permutations.ts

```diff
@@ -0,0 +1,95 @@
+import { PluggableInventoryItemDefinition } from 'app/inventory/item-types';
+import { DestinyEnergyType } from 'bungie-api-ts/destiny2';
+import { ProcessMod } from './process-worker/types';
+
+export function generateProcessModPermutations(mods: (ProcessMod | null)[]) {
+  // Creates a string from the mod permutation containing the unique properties
+  // that we care about, so we can reudce to the minimum number of permutations.
+  // If two different mods that fit in the same socket have the same energy tyoe
+  // and cost, they are identical from the mod assignment perspective.
+  // This works because we check to see if we have already recorded this string
+  // in heaps algorithm before we add the permutation to the result.
+  const createPermutationKey = (permutation: (ProcessMod | null)[]) =>
+    permutation
+      .map((mod) => {
+        if (mod) {
+          const energyType = mod.energy?.type || DestinyEnergyType.Any;
+          const energyCost = mod.energy?.val || 0;
+          return `${energyType}${energyCost}${mod.plugCategoryHash}`;
+        }
+      })
+      .filter(Boolean)
+      .join(',');
+  return generatePermutationsOfFive(mods, createPermutationKey);
```

Comment:
@bhollis rather than passing the string function around like before I have abstracted it all away into this file to try and isolate it. Does that comment make sense this time?

Reply:
tyoes aside it looks good

Reply:
Lol I read that as typo's until just now